### PR TITLE
sql: clean up session shutdown in case of panics + misc cleanups

### PR DESCRIPTION
**Pull request by knz**
This patch does two things:

- ensures that sessions can shut down cleanly in case of panics
- avoids re-stringifying statements to generate erorr messages.

(This is a "light" version of  #15063 which does not  recover from panics)

### CODE REVIEW

**Comment by dt** [comment]
nit: commit message refers to SQL "client".

I find the distinction between which returned errors are fatal and which aren't a little cumbersome / hard to follow, but you've thought about this more than I have so if you think this is the best / only way to express that, I'll take your word for it.

I wonder if, where we're worried about the session internal state being invalid, we should mark the session as known-bad?

---

Reviewed 4 of 8 files at r1.
Review status: 4 of 8 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 490 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-KitiX5u_wxAfH2D6HXC:-KitiX5u_wxAfH2D6HXD:bw29ls9) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/executor.go#L490)):*
> ```Go
> // If the returned error is non-nil, the session cannot be used any
> // more and must be discarded.
> // TODO(knz) use this fatal error return to tell the
> ```

merge-skew alert: this will fail your new TODO linter :p

---

*[pkg/sql/executor.go, line 1268 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kits3OrEeBxVvTjm4B_:-Kits3OrEeBxVvTjm4Ba:b-vjtmq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/executor.go#L1268)):*
> ```Go
> 			qArgs[idx] = typedExpr
> 		}
> 		results, err := e.ExecuteStatements(session, prepared.Query, &parser.PlaceholderInfo{Values: qArgs, Types: prepared.SQLTypes})
> ```

nit: long line.

---

*[pkg/sql/prepare.go, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-KithSw_LsscxGtHnrAP:-KithSw_LsscxGtHnrAQ:b-wcecoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/prepare.go#L75)):*
> ```Go
> func (ps PreparedStatements) New(
> 	e *Executor, name, query string, placeholderHints parser.PlaceholderTypes,
> ) (*PreparedStatement, bool, error) {
> ```

comment (or named return)?

---

*[pkg/sql/mon/mem_usage.go, line 337 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-KitgVdqQ2Z7zpaOTqk_:-KitgVdrUqpQa6AqFsHq:b-gbznnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/mon/mem_usage.go#L337)):*
> ```Go
> }
> 
> func (mm *MemoryMonitor) doStop(ctx context.Context, check bool) {
> ```

nit: comment or, better yet, descriptive name e.g. `assertEmpty`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-:-Kithv6f0PBh69P2ihll:b-5lksjq)*

**Comment by knz** [comment]
@dt TFYR
@jordanlewis can you also have a look since I touched what we talked about

**Comment by jordanlewis** [comment]
The updates to prepare :lgtm:

---

Reviewed 8 of 8 files at r3, 3 of 3 files at r4, 2 of 2 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 421 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy0lu-GZJHlaG7PMwV:-Kjy0lu-GZJHlaG7PMwW:bzftu3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/071e7a8fba241a0291a511b362f1042431026573/pkg/sql/executor.go#L421)):*
> ```Go
> // nil if there are no results).
> func (e *Executor) Prepare(
> 	stmt parser.Statement, stmtStr string, session *Session, pinfo parser.PlaceholderTypes,
> ```

Sweet!

---

*[pkg/sql/executor.go, line 446 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy0ZLR6LYuSYUoBJZA:-Kjy0ZLR6LYuSYUoBJZB:b-40frdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/071e7a8fba241a0291a511b362f1042431026573/pkg/sql/executor.go#L446)):*
> ```Go
> 	}
> 
> 	return e.doPrepare(stmt, stmtStr, session, pinfo, txn)
> ```

Why is this not inlined?

---

*[pkg/sql/prepare.go, line 106 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy0OMaDCHXWofNJwkO:-Kjy0OMb-F22RNnQ3pen:b-dkk09r) ([raw file](https://github.com/cockroachdb/cockroach/blob/071e7a8fba241a0291a511b362f1042431026573/pkg/sql/prepare.go#L106)):*
> ```Go
> // inferring placeholder types.
> // If the 2nd return value is true, then the session is trashed and
> // cannot be used any more.
> ```

s/true/non-nil/?

---

*[pkg/sql/session.go, line 403 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy-Y9ukXVdhwK5oD_a:-Kjy-Y9ukXVdhwK5oD_b:bmrc58n) ([raw file](https://github.com/cockroachdb/cockroach/blob/3687a17e8ee8f18482bec960e934501d74be8a9e/pkg/sql/session.go#L403)):*
> ```Go
> 
> // EmergencyClose is an edulcorated version of Finish() which is less picky
> // about the current state of the Session.
> ```

This word again! But more importantly, this doesn't help me understand what this thing is doing or where it should be used. Is this a `Finish()` replacement?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-:-Kjy0rsYSWtHGzQLNrRw:bxh90e8)*

**Comment by knz** [comment]
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/executor.go, line 490 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-KitiX5u_wxAfH2D6HXC:-KjxT-IdqtXraTBKDCd_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/executor.go#L490)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

merge-skew alert: this will fail your new TODO linter :p
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1268 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kits3OrEeBxVvTjm4B_:-KjxT-egGtB9cQH_kBDs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/executor.go#L1268)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: long line.
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 446 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy0ZLR6LYuSYUoBJZA:-KkBG4A0eC3wYvMThQbs:bg8z0hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/071e7a8fba241a0291a511b362f1042431026573/pkg/sql/executor.go#L446)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why is this not inlined?
</blockquote></details>

This was an artifact from an earlier iteration that was doing something different. Sorry for that. Done.

---

*[pkg/sql/prepare.go, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-KithSw_LsscxGtHnrAP:-KkBGNbBZqA-lgr5-p_q:b-oh3ezv) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/prepare.go#L75)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

comment (or named return)?
</blockquote></details>

I don't understand?

---

*[pkg/sql/prepare.go, line 106 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy0OMaDCHXWofNJwkO:-KkBGI2rNVajNv1phY_y:bafs8m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/071e7a8fba241a0291a511b362f1042431026573/pkg/sql/prepare.go#L106)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/true/non-nil/?
</blockquote></details>

This comment shouldn't have been there in the first place. removed.

---

*[pkg/sql/session.go, line 403 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-Kjy-Y9ukXVdhwK5oD_a:-KkBHzzRD_pBdlEUEJHO:bjk759z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3687a17e8ee8f18482bec960e934501d74be8a9e/pkg/sql/session.go#L403)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This word again! But more importantly, this doesn't help me understand what this thing is doing or where it should be used. Is this a `Finish()` replacement?
</blockquote></details>

Yeah sorry for not rewriting this comment earlier. Done.

---

*[pkg/sql/mon/mem_usage.go, line 337 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358#-KitgVdqQ2Z7zpaOTqk_:-KkBI3F8xCupVxG6pQfj:bofczdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e22a6766778c044ede9a173621b6f4951fa67c/pkg/sql/mon/mem_usage.go#L337)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: comment or, better yet, descriptive name e.g. `assertEmpty`?
</blockquote></details>

I do not understand what "assertEmpty" would mean and what you think is missing here. Can you suggest which comment you would like to read here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358)*

**Comment by jordanlewis** [comment]
Reviewed 5 of 5 files at r7, 3 of 3 files at r8, 2 of 2 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15358)*

**Comment by knz** [comment]
@dt do you want me to adjust anything before this merges?
